---
title: "cluster_eisa_overlap"
author: "Neelanjan Mukherjee"
date: "10/2/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(tidyverse)
library(viridis)
library(pheatmap)
library(ggplot2)
library(ggthemes)
library(ggrepel)
library(scales)
library(limma)
library(reshape2)
library(msigdbr)
library(GeneOverlap)
```


```{r eisa by clusters}

# stability per time point
res_full <- read_tsv(here("fig4_PTR-H295R", "output", "eisa_call.csv.gz"))

# groups of genes
delta_up <- res_full %>% filter(FDR <= 0.01) %>% filter(logFC > 0) %>% pull(gene) %>% unique()
delta_down <- res_full %>% filter(FDR <= 0.01) %>% filter(logFC < 0) %>% pull(gene) %>% unique()
delta_mix <- intersect(delta_down, delta_up)
delta_non <- setdiff(res_full$gene %>% unique(), c(delta_up, delta_down))

# define groups in the table
res_full2 <- res_full %>% mutate(eisa = case_when(
  gene %in% delta_mix ~ "mix",
  gene %in% delta_up ~ "up",
  gene %in% delta_down ~ "down",
  gene %in% delta_non ~ "insig",
  TRUE ~ "NA"
)) %>% select(gene, eisa) %>% distinct()


load(here("data","RNAseq","clustGenes.R"))
clust_gene <- split(myClust_annotated$symbol,myClust_annotated$cluster)

eisa_gene <- split(res_full2$gene, res_full2$eisa)
eisa_gene <- eisa_gene[- 3]

gom.eisa_clust <- newGOM(eisa_gene, clust_gene, genome.size = nrow(res_full2))

mymat <- getMatrix(gom.eisa_clust, "odds")
ceil <- 10

mymat[mymat > ceil] <- ceil
mymat <- mymat[c("up","insig","down"),sort(colnames(mymat))]


pheatmap::pheatmap(t(mymat), cluster_cols = F, cluster_rows = F, color = viridis::viridis(10, option = "E", direction = 1), border_color = "black", cellwidth = 15, cellheight = 15, fontsize = 10, filename = here("fig4_PTR-H295R","plots","cluster_eisa_overlap.pdf"), silent = T)
dev.off()


```

## Examples

```{r}
anno <- read_csv(here("fig4_PTR-H295R","output","eisa_cluster_genes.csv.gz"))



myTxi_H295R <- readRDS(here("data", "RNAseq", "myTxi_H295R.rds"))


# myG <- "STAR"

getvalues <- function(myG) {
preG <- myTxi_H295R$counts$precursor %>%  filter(symbol==myG) %>% select_if(is.numeric) %>% reshape2::melt() %>% mutate(cat=rep("pre_mRNA",24))

preG <- separate(data = preG, col = variable, sep = "_", into = c("stim","time","rep"))

preMean <- preG %>% filter(time==0) %>% pull(value) %>% mean()

preG$value <- preG$value/preMean


matG <- myTxi_H295R$counts$matureGene %>% filter(symbol==myG) %>% select_if(is.numeric) %>% reshape2::melt() %>% mutate(cat=rep("mRNA",24))

matG <- separate(data = matG, col = variable, sep = "_", into = c("stim","time","rep"))

matMean <- matG %>% filter(time==0) %>% pull(value) %>% mean()

matG$value <- matG$value/matMean

both_mat <- bind_rows(matG,preG)
colnames(both_mat)[4] <- "lfc"
both_mat$time <- as.numeric(both_mat$time)

return(both_mat)
}


### aw3
p_STAR <- ggpubr::ggline(data = getvalues("STAR"), x = "time", y = "lfc", group = "cat", numeric.x.axis = T, add = "mean", error.plot = "errorbar", color = "cat", palette = c("red","black"), title = "STAR")


p_GPRC5A <- ggpubr::ggline(data = getvalues("GPRC5A"), x = "time", y = "lfc", group = "cat", numeric.x.axis = T, add = "mean", error.plot = "errorbar", color = "cat", palette = c("red","black"), title = "GPRC5A")



### as3
p_MC2R <- ggpubr::ggline(data = getvalues("MC2R"), x = "time", y = "lfc", group = "cat", numeric.x.axis = T, add = "mean", error.plot = "errorbar", color = "cat", palette = c("red","black"), title = "MC2R")


p_CITED2 <-ggpubr::ggline(data = getvalues("CITED2"), x = "time", y = "lfc", group = "cat", numeric.x.axis = T, add = "mean", error.plot = "errorbar", color = "cat", palette = c("red","black"), title = "CITED2")


p_all <- cowplot::plot_grid(
  p_MC2R + theme(legend.position="none"),
  p_CITED2 + theme(legend.position="none"),
  p_STAR + theme(legend.position="none"),
  p_GPRC5A + theme(legend.position="none")
  )

legend_b <- cowplot::get_legend(
  p_STAR +
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

# add the legend underneath the row we made earlier. Give it 10%
# of the height of one plot (via rel_heights).
p_all_bottom <- cowplot::plot_grid(p_all, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(plot = p_all_bottom, filename = here("fig4_PTR-H295R","plots","precursor_mature_examples.pdf"), width =6, height = 4)

```


