---
title: "running cword"
author: "Rui Fu"
date: ""
output: html_document
---

```{r setup}
# Import R packages that we will need in the analysis
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
ggplot2::theme_set(ggplot2::theme_classic())
library(tidyverse)
library(here)
```

```{r load}
load(here("data", "RNAseq", "clustGenes.R"))
myTxi_H295R <- readRDS(here("data", "RNAseq", "myTxi_H295R.rds"))
clustList <- myClust_annotated %>% pull(symbol) %>% split(myClust_annotated$cluster)
load(here("data", "accessories", "clusterCols"))
```

```{r}
DE_utr3s_seqs <- readRDS(here("fig4_pTR-H295R", "output", "DE_utr3s_seqs.rds"))
utr3s_seqs <- readRDS(here("fig4_pTR-H295R", "output", "utr3s_seqs.rds"))
utr3s_seqs <- utr3s_seqs %>% left_join(myClust_annotated %>% select(name = symbol, cluster) %>% mutate(cluster = as.character(cluster))) %>%
  replace_na(list(cluster = "nonDE"))

res_full <- read_tsv(here("fig4_PTR-H295R", "output", "eisa_call.csv.gz"))
delta_up <- res_full %>% filter(FDR <= 0.05) %>% filter(logFC > 0) %>% pull(gene) %>% unique()
delta_down <- res_full %>% filter(FDR <= 0.05) %>% filter(logFC < 0) %>% pull(gene) %>% unique()
delta_mix <- intersect(delta_down, delta_up)
delta_non <- setdiff(res_full$gene %>% unique(), c(delta_up, delta_down))
allseqs <- bind_rows(DE_utr3s_seqs, utr3s_seqs %>% filter(cluster == "nonDE")) %>% 
  mutate(cluster = case_when(
    name %in% delta_down ~ "down",
    name %in% delta_up ~ "up",
    name %in% delta_mix ~ "mix",
    name %in% delta_non ~ "no",
    TRUE ~ "nonDE"
  ))
```

```{r prep_fa}
write_fasta <- function(df, out_path,
                        header_col = "name",
                        seq_col = "seq",
                        gz = FALSE,
                        cutoff = 20, 
                        remove = "N") {

  stopifnot(all(c(header_col, seq_col) %in% colnames(df)))
  
  hc <- df[[header_col]]
  sc <-  df[[seq_col]]
  
  cutfrom <- str_length(sc) < cutoff
  hc2 <- hc[!cutfrom]
  sc2 <- sc[!cutfrom]
  
  if (!(is.null(remove))) {
  sc2 <- sc2 %>% str_remove_all(paste0("^", remove, "+")) %>% str_remove_all(paste0(remove, "+", "$"))
  removefrom <- str_detect(sc2, remove)
  hc3 <- hc2[!removefrom]
  sc3 <- sc2[!removefrom]
  } else {
    sc3 <- sc2
    hc3 <- hc2
  }

  res <- paste0(">", hc3, "\n", sc3)
  write_lines(res, path.expand(out_path))

  if(gz) R.utils::gzip(path.expand(out_path), remove = T, overwrite = T)
}

write_fasta(allseqs, here("fig4_PTR-H295R", "tmp", "allseqs.fa"))
```

```{r prep_rnk}
res_full <- read_tsv(here("fig4_PTR-H295R", "output", "eisa_call.csv.gz"))
delta_up <- res_full %>% filter(FDR <= 0.01) %>% filter(logFC > 0) %>% pull(gene) %>% unique()
delta_down <- res_full %>% filter(FDR <= 0.01) %>% filter(logFC < 0) %>% pull(gene) %>% unique()
res_up_max <- res_full %>% group_by(gene) %>% 
  filter(gene %in% delta_up) %>% 
  summarise(fc = max(logFC)) %>% 
  ungroup() %>% 
  arrange(desc(fc))
write_tsv(res_up_max, here("fig4_PTR-H295R", "tmp", "eisa_up.rnk"), col_names = FALSE)
res_down_max <- res_full %>% group_by(gene) %>% 
  filter(gene %in% delta_down) %>% 
  summarise(fc = min(logFC)) %>% 
  ungroup() %>% 
  arrange(fc) %>% 
  mutate(fc = -fc)
write_tsv(res_down_max, here("fig4_PTR-H295R", "tmp", "eisa_down.rnk"), col_names = FALSE)
write_tsv(bind_rows(res_up_max, res_down_max %>% mutate(fc = -fc)) %>% arrange(fc),
          here("fig4_PTR-H295R", "tmp", "eisa.rnk"), col_names = FALSE)

delta_up_2 <- res_full %>% filter(time >= 2, FDR <= 0.01) %>% filter(logFC > 0) %>% pull(gene) %>% unique()
delta_down_2 <- res_full %>% filter(time >= 2, FDR <= 0.01) %>% filter(logFC < 0) %>% pull(gene) %>% unique()
res_up_max <- res_full %>% group_by(gene) %>% 
  filter(gene %in% delta_up_2) %>% 
  summarise(fc = max(logFC)) %>% 
  ungroup() %>% 
  arrange(desc(fc))
write_tsv(res_up_max, here("fig4_PTR-H295R", "tmp", "eisa_up2.rnk"), col_names = FALSE)
res_down_max <- res_full %>% group_by(gene) %>% 
  filter(gene %in% delta_down_2) %>% 
  summarise(fc = min(logFC)) %>% 
  ungroup() %>% 
  arrange(fc) %>% 
  mutate(fc = -fc)
write_tsv(res_down_max, here("fig4_PTR-H295R", "tmp", "eisa_down2.rnk"), col_names = FALSE)

res_up_fdr <- res_full %>% group_by(gene) %>% 
  filter(gene %in%  delta_up) %>% 
  summarise(FDR = min(FDR)) %>% 
  ungroup() %>% 
  arrange(desc(FDR))
write_tsv(res_down_fdr, here("fig4_PTR-H295R", "tmp", "eisa_up_fdr.rnk"), col_names = FALSE)
```


```{bash, eval = FALSE}
# jruby -S gem install progressbar -v 0.9.0

#cwords_mkdb -w 4,5 -s /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/allseqs.fa -t 1 --bg 2

#cwords --db allseqs_bg2 -w 4,5 -p 100 -r /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/eisa.rnk

cwords -u 0 -w 5,6 -x -r /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/eisa.rnk -s /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/allseqs.fa -p 100 -c obs
```

```{r prep_rnk_decay}
rates <- read_tsv(file = here("data/labeling/all_rates.tsv.gz"))

write_tsv(rates %>% select(gene_name, degradation_tpt_0) %>% arrange(desc(degradation_tpt_0)), here("fig4_PTR-H295R", "tmp", "decay0.rnk"), col_names = FALSE)
```

```{r prep_aw3}
d1 <- res_full %>% inner_join(myClust_annotated, by = c("gene" = "symbol")) %>% 
  filter(cluster == "aw3") %>% 
  filter(FDR <= 0.01, logFC < 0) %>% 
  group_by(gene) %>% 
  summarize(logFC = min(logFC)) %>% 
  arrange(logFC) 
d2 <- res_full %>% inner_join(myClust_annotated, by = c("gene" = "symbol")) %>% 
  filter(cluster == "aw3") %>% 
  group_by(gene) %>% 
  summarize(logFC = min(logFC), insig = sum(FDR > 0.01)) %>% 
  filter(insig == 11) %>% 
  mutate(logFC = 0)
bind_rows(d1, d2) %>% 
  write_tsv(here("fig4_PTR-H295R", "tmp", "aw3.rnk"), col_names = FALSE)
```


```{bash, eval = FALSE}
# jruby -S gem install progressbar -v 0.9.0

#cwords_mkdb -w 4,5 -s /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/allseqs.fa -t 1 --bg 2

#cwords --db allseqs_bg2 -w 4,5 -p 100 -r /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/eisa.rnk

cwords -u 0 -w 5,6 -x -r /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/decay0.rnk -s /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/allseqs.fa -p 100 -c obs

cwords -u 0 -w 5,6 -x -r /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/aw3.rnk -s /Users/rf/2020_PTR_steroidogenesis_paper/fig4_PTR-H295R/tmp/allseqs.fa -p 100 -c obs
```
